File: nodejs/matrix_mult_test.js
onst mathjs = require('mathjs');

/**
 * 使用mathjs库进行矩阵乘法测试
 */
function matrixMultiplyMathjs(size) {
    // 创建两个随机矩阵
    const matrixA = mathjs.random([size, size]);
    const matrixB = mathjs.random([size, size]);
    
    // 执行乘法并计时
    const start = process.hrtime.bigint();
    const result = mathjs.multiply(matrixA, matrixB);
    const end = process.hrtime.bigint();
    
    // 转换为秒
    return Number(end - start) / 1e9;
}

/**
 * 手动实现矩阵乘法
 */
function matrixMultiplyManual(size) {
    // 创建两个随机矩阵
    const matrixA = Array(size).fill().map(() => 
        Array(size).fill().map(() => Math.random()));
    const matrixB = Array(size).fill().map(() => 
        Array(size).fill().map(() => Math.random()));
    
    // 初始化结果矩阵
    const result = Array(size).fill().map(() => Array(size).fill(0));
    
    // 计时开始
    const start = process.hrtime.bigint();
    
    // 执行矩阵乘法
    for (let i = 0; i < size; i++) {
        for (let j = 0; j < size; j++) {
            for (let k = 0; k < size; k++) {
                result[i][j] += matrixA[i][k] * matrixB[k][j];
            }
        }
    }
    
    // 计时结束
    const end = process.hrtime.bigint();
    
    // 转换为秒
    return Number(end - start) / 1e9;
}

function main() {
    // 从命令行获取参数或使用默认值
    const size = process.argv.length > 2 ? parseInt(process.argv[2]) : 1000;
    
    console.log(`执行 ${size}x${size} 矩阵乘法测试`);
    
    // mathjs实现
    const mathjsTime = matrixMultiplyMathjs(size);
    console.log(`mathjs实现: 用时 = ${mathjsTime.toFixed(6)} 秒`);
    
    // 如果矩阵不太大，可以测试手动实现
    if (size <= 200) {
        const manualTime = matrixMultiplyManual(size);
        console.log(`手动实现: 用时 = ${manualTime.toFixed(6)} 秒`);
    } else {
        console.log("矩阵太大，跳过手动实现测试");
    }
}

main();

File: python/matrix_mult_test.py
#!/usr/bin/env python3
import time
import numpy as np
import sys

def matrix_multiply_numpy(size):
    """使用NumPy执行矩阵乘法"""
    # 创建两个随机矩阵
    matrix_a = np.random.rand(size, size)
    matrix_b = np.random.rand(size, size)
    
    start = time.time()
    result = np.matmul(matrix_a, matrix_b)
    end = time.time()
    
    return end - start

def matrix_multiply_manual(size):
    """手动实现矩阵乘法（性能较差，仅作为比较）"""
    # 创建两个随机矩阵
    matrix_a = [[np.random.random() for _ in range(size)] for _ in range(size)]
    matrix_b = [[np.random.random() for _ in range(size)] for _ in range(size)]
    
    # 初始化结果矩阵
    result = [[0 for _ in range(size)] for _ in range(size)]
    
    start = time.time()
    # 执行矩阵乘法
    for i in range(size):
        for j in range(size):
            for k in range(size):
                result[i][j] += matrix_a[i][k] * matrix_b[k][j]
    end = time.time()
    
    return end - start

if __name__ == "__main__":
    size = 1000
    if len(sys.argv) > 1:
        size = int(sys.argv[1])
    
    print(f"执行 {size}x{size} 矩阵乘法测试")
    
    # NumPy实现
    numpy_time = matrix_multiply_numpy(size)
    print(f"NumPy实现: 用时 = {numpy_time:.6f} 秒")
    
    # 如果矩阵不太大，可以测试手动实现
    if size <= 200:
        manual_time = matrix_multiply_manual(size)
        print(f"手动实现: 用时 = {manual_time:.6f} 秒")
    else:
        print("矩阵太大，跳过手动实现测试")